site/docs/skylark/aspects.md

```diff
@@ -116,85 +150,242 @@ functions. They return [providers](rules.md#providers), can generate
+    attrs = {
+      "extension" : attr.string(values = ["*", ".h", ".cc"]),
+    }
+)
+
+def _file_count_rule_impl(ctx):
+  for dep in ctx.attr.deps:
+    print(dep[FileCount].count)
+  
+file_count_rule = rule(
+    implementation =  _file_count_rule_impl,
+    attrs = {
+        'deps' : attr.label_list(aspects = [file_count_aspect]),
+        'extension' : attr.string(default = "*"),
+    },
+)
 ```
 
-The implementation function can access the attributes of the target rule via
-[`ctx.rule.attr`](lib/ctx.html#rule). It can examine providers that are
-provided  by the target to which it is applied (via the `target` argument).
+BUILD.bazel file:
+```python
+load('//file_count.bzl', 'file_count_rule')
+
+cc_library(
+    name = 'lib',
+    srcs = [
+        'lib.h',
+        'lib.cc',
+    ],
+)
+
+cc_binary(
+    name = 'app',
+    srcs = [
+        'app.h',
+        'app.cc',
+        'main.cc',
+    ],
+    deps = ['lib'],
+)
+
+file_count_rule(
+  name = 'file_count',
+  deps = ['app'],
+  extension = ".h",
+)
+```
+
+### Aspect definition
+
+```python
+file_count_aspect = aspect(implementation = _file_count_aspect_impl,
+    attr_aspects = ["deps"],
+    attrs = {
+      "extension" : attr.string(values = ["*", ".h", ".cc"]),
+    }
+)
+```
+
+In this example we are again propagating the aspect via the ``deps`` attribute.
+
+``attrs`` defines a set of attributes for an aspect. Public aspect attributes 
+are of type ``string`` and are called parameters. Parameters must have a``values`` 
+attribute specified on them. In this case we have a parameter called ``extension`` 
+that is allowed to have "\*", ".h", or ".cc" as a value. 
+
+Parameter values for the aspect are taken from the string attribute with the same 
+name of the rule requesting the aspect (see the definition of ``file_count_rule``). 
+Aspects with parameters cannot be used via the command line because there is no
+syntax to define the parameters.
+
+Aspects are also allowed to have private attributes of types ``label`` or 
+``label_list``. Private label attributes can be used to specify dependencies on 
+tools or libraries that are needed for actions generated by aspects. There is not
+a private attribute defined in this example, but the following code snippet
+demonstrates how you could pass in a tool to a aspect:
+
+```python
+...
+    attrs = {
+      "_protoc" : attr.label(
+          default = Label("//tools:protoc"),
+          executable = True
+      )
+    }
+...
+```
+
+### Aspect implementation
+
+```python
+FileCount = provider(
+    fields = {
+        'count' : "number of files"
+    }
+)
+
+def _file_count_aspect_impl(target, ctx):
+    count = 0
+    # Make sure the rule has a srcs attribute.
+    if hasattr(ctx.rule.attr, "srcs"):
+        # Iterate through the sources counting files
+        for src in ctx.rule.attr.srcs:
+            for f in src.files:
+                if ctx.attr.extension == "*" or f.path.endswith(ctx.attr.extension):
```

Comment:
You can also use `.extension` attribute on `f` (if it's a File instance)

Reply:
Fixed.